File: js/solver.js

(function() {

    "use strict";

    // [[1, 2, 3], [1, 2, 3]].flatten() gives [1, 2, 3, 1, 2, 3]
    Array.prototype.flatten = () => [].concat.apply([], this);

    var DIGITS = [1, 2, 3, 4, 5, 6, 7, 8, 9];

    // These are a couple of useful map functions, pulling them up here
    // speeds things up.
    var valueOfCell = (cell => cell.value),
        possibleValuesOfCell = (cell => cell.possibleValues),
        isUnsolved = (cell => cell.value === 0);

    class Solver {

        constructor(grid) {
            this.grid = grid;
            this.guesses = 0;
            this._solvedCells = [];
        }

        solve() {
            // work out what the set of possible values is for each unsolve cell.
            this._initPossibleValues();
            // if there are any with only one possible value, set it.
            this._findCellsWithOnePossibleValue();
            // find any cells that have a possible value that doesn't occur
            // anywhere else in the column, row or subgrid it's in
            this._findUniqueValuesInUnits();
            if (!this.grid.isSolved()) {
                // this is a more tricky puzzle, so start searching for a solution.
                this._search();
            }
        }

        _search() {
            // pick the cell with least possible values (more chance of guessing correctly)
            var cell = (this.grid.unsolved().sort((x, y) => {
                var xVal = x.possibleValues.length * 100 + (x.row + x.col),
                    yVal = y.possibleValues.length * 100 + (y.row + y.col);

                return xVal - yVal;
            }))[0];

            for (let value of cell.possibleValues) {
                // remember how many cells we had solved before we begin incase
                // we need to unwind
                let numSolved = this._solvedCells.length;
                this.guesses += 1;

                try {
                    this._setValueForCell(cell, value);
                    if (!this.grid.isSolved()) {
                        // no luck, keep looking...
                        this._search();                      
                    }
                } catch (inconsistency) {
                    // here's the back tracking part, we've ended up in a position where we
                    // can't progress, so before we try another value, undo all the values
                    // we set since the last guess.   
                    let resetPossibilities = [];   
                    this._solvedCells.splice(numSolved, this._solvedCells.length - numSolved)
                                     .forEach(cell => { 
                                        cell.value = 0;
                                        resetPossibilities = resetPossibilities.concat(this.grid.peers(cell));
                                     }, this);

                    this._initPossibleValues(new Set(resetPossibilities.filter(isUnsolved)));
                }
            }
            if (!this.grid.isSolved()) {
                // If we get here then we're also stuck since we haven't found a solution despite trying
                // all possible values for a cell.
                throw "Tried all values for this cell  [" + cell.row + ", " + cell.col + "]" + cell.possibleValues;
            }
        }

        _initPossibleValues(cells) {
            (cells || this.grid.unsolved()).forEach((cell) => {
                let peerValues = this.grid.peers(cell).map(valueOfCell),
                    possibleValues = DIGITS.filter(d => peerValues.indexOf(d) === -1);
                    cell.possibleValues = possibleValues;
            }, this);
        }

        _removeValueFromPeers(cell) {
            // Summary:
            //  Remove the value of cell from the possible values of 
            //  it's peers.
            this.grid.peers(cell)
                    .filter(isUnsolved)
                    .forEach((p) => { 
                        var idx = p.possibleValues.indexOf(cell.value);
                        if (idx !== -1) {
                            p.possibleValues.splice(idx, 1);
                        }

                        if (p.possibleValues.length === 0) {
                            throw "No possible values for cell ["+p.row+", "+p.col+"] "+p.value;
                        }
                    });
        }

        _setValueForCell(cell, value) {
            var peers = this.grid.peers(cell);
           
            if (peers.some(x => x.value === value)) {
                throw "Tried to set a value that already exists in peers";
            }
            
            cell.value = value;
            this._solvedCells.push(cell);
            this._removeValueFromPeers(cell);
            this._findCellsWithOnePossibleValue(peers);
            this._findUniqueValuesInUnits(cell);
        }

        _findCellsWithOnePossibleValue(cells) {
            cells = (cells || this.grid.unsolved());
            cells.forEach(cell => {
                if (cell.value === 0 && cell.possibleValues.length === 1) {
                    this._setValueForCell(cell, cell.possibleValues[0]);
                }
            }, this);
        }

        _findUniqueValuesInUnits(cell) {

            if (cell) {
                [this.grid.sameSubGridAs(cell).flatten(),
                 this.grid.sameColAs(cell),
                 this.grid.sameRowAs(cell)].forEach(this._findUniquePossibiltyInUnit, this);
            } else {
                let subGrids = this.grid.subgrids().map(sg => sg.flatten());
            
                for (let units of [subGrids, this.grid.columns(), this.grid.rows]) {
                    for (let unit of units) {
                        this._findUniquePossibiltyInUnit(unit);
                    }
                }
            }
        }

        _findUniquePossibiltyInUnit(unit) {
            let unsolved = unit.filter(isUnsolved);
            unsolved.forEach(unsolvedCell => {
                var unique,
                    otherCellsPossValues = unit
                        .filter(c => c !== unsolvedCell)
                        .map(possibleValuesOfCell)
                        .flatten();
                        //.reduce((a, b) => a.concat(b));  
               
                unique = unsolvedCell.possibleValues.filter(x => otherCellsPossValues.indexOf(x) === -1);
                if (unique.length === 1) {
                    this._setValueForCell(unsolvedCell, unique[0]);
                }
            }, this);
        }
    }

    module.exports = Solver;

}());

File: ruby/solver.rb
require 'set'

class Solver

    attr_accessor :guesses

    def initialize(grid)
        @grid = grid
        @guesses = 0
        @solved_cells = []
        @digits = (1..9).to_a
    end

    def solve
        # work out what the set of possible values is for each unsolved cell.
        init_possible_values(nil)
        # if there are any with only one possible value, set it.
        find_cells_with_one_possible_value(@grid.unsolved)
        # find any cells that have a possible value that doesn't occur
        # anywhere else in the column, row or subgrid it's in
        find_unique_values_in_units(nil)
        unless @grid.is_solved?
            # this is a more tricky puzzle, so start searching for a solution.
            search
        end
    end

    def search
        # pick the cell with least possible values (more chance of guessing correctly)
        cell = @grid.unsolved.min { |x, y| 
            xval = x.possible_values.length * 100 + (x.row + x.col) 
            yval = y.possible_values.length * 100 + (y.row + y.col) 

            xval <=> yval
        }

        cell.possible_values.each do |value|        
            # remember how many cells we had solved before we begin incase
            # we need to unwind
            num_solved = @solved_cells.length
            @guesses += 1

            begin
                set_value_for_cell(cell, value)
                unless @grid.is_solved?
                    # no luck, keep looking...
                    search   
                end                   
                
            rescue Exception => inconsistency
                reset_possibilities = []
                @solved_cells.slice!(num_solved, @solved_cells.length - num_solved)
                                    .each do |cell|  
                                        cell.value = 0;
                                        reset_possibilities << cell
                                        reset_possibilities.concat(@grid.peers(cell).to_a)
                                    end

                init_possible_values(reset_possibilities.select { |x| x.value == 0 }.to_set)
            end
        end

        unless @grid.is_solved?
            # If we get here then we're also stuck since we haven't found a solution despite trying
            # all possible values for a cell.
            raise "Tried all values for this cell  [#{cell.row},  #{cell.col}]"
        end
    end

    def init_possible_values(cells)
        ''' 
            Initialise the possible values for the provided list of cells or
            all the unsolved cells in the grid if no list was provided.

            To do this we collect the "peers" for each cell (cells not marked . for the cell c):

            x x x | . . . | . . .
            5 c x | x x 2 | x 9 x
            x x 3 | . . . | . . .
            ------+-------+------
            . x . | . . . | . . .
            . x . | . . . | . . .
            . x . | . . . | . . .
            ------+-------+------
            . x . | . . . | . . .
            . 7 . | . . . | . . .
            . x . | . . . | . . .

            Remove from the peers any unsolved cells, then exclude from the list 1..9 any 
            numbers already present in the list of solved peers. e.g. in the above grid assuming
            that any cell containing an x or a number is a peer of c and that the cells containing
            the numbers are solved then the possible values for "c" are:

            [1, 2, 3, 4, 5, 6, 7, 8, 9] - [5, 3, 2, 9, 7] = [8, 1, 4, 6]

        '''
        if cells.nil?
            cells = @grid.unsolved
        end

        cells.each do |cell| 
            # get the value of all solved peers
            peer_values = @grid.peers(cell).select {|x| x.value != 0 }.map { |x| x.value }
            # remove any numbers from 1..9 that exist in that values.
            cell.possible_values = @digits.select { |d| !peer_values.include? d }
        end
    end

    def remove_value_from_peers(cell)
        '''
            Remove the value of the given cell from the possible values
            of any unsolved peer cells.

            If this results in a cell having no possible values, then 
            raise an exception.
        '''
        @grid.peers(cell)
            .select { |x| x.value == 0 }
            .each do |p| 
                p.possible_values.delete_if { |v| v == cell.value }
                if p.possible_values.length == 0
                    raise "No possible values for cell [#{p.row} , #{p.col}], having set #{cell.row}, #{cell.col}"
                end
            end
    end

    def set_value_for_cell(cell, value)
        peers = @grid.peers(cell)
           
        if peers.any? { |x| x.value === value }
            raise "Tried to set a value for #{cell.row}, #{cell.col} that already exists in peers #{value}"
        end
            
        cell.value = value
        cell.possible_values = []
        @solved_cells << cell
        remove_value_from_peers(cell)
        find_cells_with_one_possible_value(peers)
        find_unique_values_in_units(cell)
    end
        
    def find_cells_with_one_possible_value(cells)
        if cells.nil?
            cells = @grid.unsolved
        end
        
        cells.each do |cell| 
            if cell.value == 0 and cell.possible_values.length == 1
                set_value_for_cell(cell, cell.possible_values[0])
            end
        end
    end

    def find_unique_values_in_units(cell) 
        unless cell.nil?
            [@grid.same_sub_grid_as(cell).flatten,
             @grid.same_col_as(cell),
             @grid.same_row_as(cell)].each { |unit| find_unique_possibilty_in_unit(unit) }
        else
            subgrids = []
            @grid.subgrids.each { |sg| subgrids << sg.flatten }
            [subgrids, @grid.columns, @grid.rows].each do |list_of_units|
                list_of_units.each { |unit| find_unique_possibilty_in_unit(unit) }
            end
        end
    end

    def find_unique_possibilty_in_unit(unit)
        unsolved = unit.select { |x| x.value == 0 }
        unsolved.each do |unsolved_cell| 
            other_cells_poss_values = unit
                    .select { |c| c != unsolved_cell and c.value == 0 }
                    .map { |c| c.possible_values }
                    .flatten
               
            unique = unsolved_cell.possible_values.select { |x| !(other_cells_poss_values.include?(x)) }

            if (unique.length === 1)
                set_value_for_cell(unsolved_cell, unique[0])         
            end
        end
    end
end


                       